File: src/main/java/org/b0n541/pmcts/mcts/Tree.java
package org.b0n541.pmcts.mcts;


import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public final class Tree {

    private static final Logger LOG = LoggerFactory.getLogger(Tree.class);

    private final TreeNode rootNode;

    public Tree(final GameState gameState) {
        rootNode = new TreeNode(gameState);
    }

    public TreeNode getRootNode() {
        return rootNode;
    }

    public long getSize() {
        return getSize(rootNode);
    }

    private static long getSize(final TreeNode node) {
        return 1 + node.getChildren().stream()
                .map(child -> getSize(child))
                .reduce(0L, Long::sum);
    }

    public void printDigraph() {
        System.out.println("======================================");
        printTreeLevel(rootNode);
        System.out.println("======================================");
    }

    private static void printTreeLevel(final TreeNode node) {
        printNodeLabel(node);
        printNodeMoves(node);

        if (!node.isLeafNode()) {
            node.getChildren().stream().filter(child -> child.getVisits() > 0).forEach(Tree::printTreeLevel);
        }
    }

    private static void printNodeLabel(final TreeNode node) {
        System.out.println(
                node.getNodeId() +
                        " [label=\"" + node.getGameState().getPlayer() + "\\n" +
                        " v=" + node.getVisits() +
                        " s=" + node.getTotalScores() + "\"" +
                        "]");
    }

    private static void printNodeMoves(final TreeNode parent) {
        parent.getMovesAndChildren().entrySet().stream()
                .filter(entry -> entry.getValue().getVisits() > 0)
                .forEach(entry -> System.out.println(
                        parent.getNodeId() +
                                " -> " + entry.getValue().getNodeId() +
                                " [label=\"" + entry.getKey().toShortString() + "\"]"));
    }

    public GameMove getBestMove() {
        long highestVisits = 0;
        GameMove bestMove = null;
        for (final TreeNode child : rootNode.getChildren()) {
            final GameMove move = child.getGameState().getLastMove();
            final long visits = child.getVisits();
            final Map<String, Double> scores = child.getTotalScores();
            LOG.info("Move {} got visits {} and scores {}", move, visits, scores);

            if (child.getVisits() > highestVisits) {
                highestVisits = visits;
                bestMove = move;
            }
        }

        LOG.info("Best move {} with highest visits {}", bestMove, highestVisits);

        return bestMove;
    }
}

File: src/main/kotlin/net/b0n541/pmcts/mcts/Tree.kt
package net.b0n541.pmcts.mcts

import org.slf4j.LoggerFactory

class Tree(gameState: GameState<GameMove>) {

    val rootNode: TreeNode = TreeNode(null, gameState)

    val size: Long
        get() = getSize(rootNode)

    fun printDigraph() {
        println("======================================")
        printTreeLevel(rootNode)
        println("======================================")
    }

    val bestMove: GameMove
        get() {
            var highestVisits: Long = 0
            var bestMove: GameMove? = null

            rootNode.children()
                .forEach {
                    val move = it.gameState.lastMove
                    val visits = it.visits
                    val scores = it.totalScores
                    LOG.info("Move {} got visits {} and scores {}", move, visits, scores)
                    if (it.visits > highestVisits) {
                        highestVisits = visits
                        bestMove = move
                    }
                }

            LOG.info("Best move {} with highest visits {}", bestMove, highestVisits)
            return bestMove!!
        }

    companion object {
        private val LOG = LoggerFactory.getLogger(Tree::class.java)
        private fun getSize(node: TreeNode): Long {
            return 1 + node.children()
                .map { getSize(it) }
                .sum()
        }

        private fun printTreeLevel(node: TreeNode) {
            printNodeLabel(node)
            printNodeMoves(node)
            if (!node.isLeafNode) {
                node.children()
                    .filter { it.visits > 0 }
                    .forEach { printTreeLevel(it) }
            }
        }

        private fun printNodeLabel(node: TreeNode) {
            println("$node.nodeId [label=\"$node.player\n v=$node.visits s=$node.totalScores\"]")
        }

        private fun printNodeMoves(parent: TreeNode) {
            parent.children()
                .filter { it.visits > 0 }
                .forEach {
                    println("$parent.nodeId -> $it.value.nodeId [label=\"$it.key.toShortString()\"]")
                }
        }
    }
}